lxd/api_cluster.go

```diff
@@ -2233,3 +2242,436 @@ func internalClusterRaftNodeDelete(d *Daemon, r *http.Request) response.Response
+				continue
+			}
+
+			dest, err := cluster.Connect(targetNode.Address, d.endpoints.NetworkCert(), d.serverCert(), r, true)
+			if err != nil {
+				return errors.Wrap(err, "Failed to connect to destination")
+			}
+
+			metadata["evacuation_progress"] = fmt.Sprintf("Starting %s", inst.Name())
+			op.UpdateMetadata(metadata)
+
+			startOp, err := dest.UpdateInstanceState(inst.Name(), api.InstanceStatePut{Action: "start"}, "")
+			if err != nil {
+				return err
+			}
+
+			err = startOp.Wait()
+			if err != nil {
+				return err
+			}
+		}
+
+		return nil
+	}
+
+	op, err := operations.OperationCreate(d.State(), "", operations.OperationClassTask, db.OperationClusterMemberEvacuate, nil, nil, run, nil, nil, r)
+	if err != nil {
+		return response.InternalError(err)
+	}
+
+	reverter.Success()
+	return operations.OperationResponse(op)
+}
+
+func restoreClusterMember(d *Daemon, r *http.Request) response.Response {
+	originName := mux.Vars(r)["name"]
+
+	var node db.NodeInfo
+	var err error
+
+	err = d.cluster.Transaction(func(tx *db.ClusterTx) error {
+		// Get the node.
+		node, err = tx.GetNodeByName(originName)
+		if err != nil {
+			return errors.Wrap(err, "Failed to get node by name")
+		}
+
+		return nil
+	})
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	if node.State == db.ClusterMemberStatePending {
+		return response.SmartError(fmt.Errorf("Cannot restore or restore a pending node"))
+	}
+
+	if node.State == db.ClusterMemberStateCreated {
+		return response.SmartError(fmt.Errorf("Node is not evacuated"))
+	}
+
+	var dbInstances []db.Instance
+
+	err = d.cluster.Transaction(func(tx *db.ClusterTx) error {
+		dbInstances, err = tx.GetInstances(*db.InstanceFilterAllInstances())
+		if err != nil {
+			return errors.Wrap(err, "Failed to get instances")
+		}
+
+		return nil
+	})
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	instances := make([]instance.Instance, 0)
+	localInstances := make([]instance.Instance, 0)
+
+	for _, dbInst := range dbInstances {
+		if dbInst.Node == node.Name {
+			inst, err := instance.LoadByProjectAndName(d.State(), dbInst.Project, dbInst.Name)
+			if err != nil {
+				return response.SmartError(errors.Wrap(err, "Failed to load instance"))
+			}
+
+			localInstances = append(localInstances, inst)
+			continue
+		}
+
+		// Only consider instances where volatile.evacuate.origin is set to the node which needs to be restored.
+		val, ok := dbInst.Config["volatile.evacuate.origin"]
+		if !ok || val != node.Name {
+			continue
+		}
+
+		inst, err := instance.LoadByProjectAndName(d.State(), dbInst.Project, dbInst.Name)
+		if err != nil {
+			return response.SmartError(errors.Wrap(err, "Failed to load instance"))
+		}
+
+		instances = append(instances, inst)
+	}
+
+	run := func(op *operations.Operation) error {
+		var source lxd.InstanceServer
+		var sourceNode db.NodeInfo
+
+		metadata := make(map[string]interface{})
+
+		// Migrate instances
+		for _, inst := range instances {
```

Comment:
Same comment as during evacuation. When restoring instances, we should do it instance by instance so that each instance is offline for the least amount of time possible.